scipy/sparse/tests/test_construct.py

```diff
@@ -385,93 +390,144 @@ def test_hstack(self):
+        assert_equal(block_array([[None, B], [C, None]]).toarray(),
                      expected)
-        assert_equal(construct.bmat([[E.T.tocsr(), B.tocsr()],
+        assert_equal(block_array([[E.T.tocsr(), B.tocsr()],
                                      [C.tocsr(), Z]]).toarray(),
                      expected)
-        assert_equal(construct.bmat([[E.T.tocsc(), B.tocsc()],
+        assert_equal(block_array([[E.T.tocsc(), B.tocsc()],
                                      [C.tocsc(), Z.tocsc()]]).toarray(),
                      expected)
 
-        expected = matrix(np.empty((0, 0)))
-        assert_equal(construct.bmat([[None, None]]).toarray(), expected)
-        assert_equal(construct.bmat([[None, D], [D, None]]).toarray(),
+        expected = np.empty((0, 0))
+        assert_equal(block_array([[None, None]]).toarray(), expected)
+        assert_equal(block_array([[None, D], [D, None]]).toarray(),
                      expected)
 
         # test bug reported in gh-5976
         expected = array([[7]])
-        assert_equal(construct.bmat([[None, D], [C, None]]).toarray(),
+        assert_equal(block_array([[None, D], [C, None]]).toarray(),
                      expected)
 
         # test failure cases
         with assert_raises(ValueError) as excinfo:
-            construct.bmat([[A], [B]])
+            block_array([[A], [B]])
         excinfo.match(r'Got blocks\[1,0\]\.shape\[1\] == 1, expected 2')
 
         with assert_raises(ValueError) as excinfo:
-            construct.bmat([[A.tocsr()], [B.tocsr()]])
+            block_array([[A.tocsr()], [B.tocsr()]])
         excinfo.match(r'incompatible dimensions for axis 1')
 
         with assert_raises(ValueError) as excinfo:
-            construct.bmat([[A.tocsc()], [B.tocsc()]])
+            block_array([[A.tocsc()], [B.tocsc()]])
         excinfo.match(r'Mismatching dimensions along axis 1: ({1, 2}|{2, 1})')
 
         with assert_raises(ValueError) as excinfo:
-            construct.bmat([[A, C]])
+            block_array([[A, C]])
         excinfo.match(r'Got blocks\[0,1\]\.shape\[0\] == 1, expected 2')
 
         with assert_raises(ValueError) as excinfo:
-            construct.bmat([[A.tocsr(), C.tocsr()]])
+            block_array([[A.tocsr(), C.tocsr()]])
         excinfo.match(r'Mismatching dimensions along axis 0: ({1, 2}|{2, 1})')
 
         with assert_raises(ValueError) as excinfo:
-            construct.bmat([[A.tocsc(), C.tocsc()]])
+            block_array([[A.tocsc(), C.tocsc()]])
         excinfo.match(r'incompatible dimensions for axis 0')
 
+    def test_block_array_return_type(self):
+        block = construct.block_array
+
+        Flist, Glist = [[1, 2],[3, 4]], [[7], [5]]
+        Fm, Gm = csr_matrix(Flist), csr_matrix(Glist)
+        assert block([[None, Flist], [Glist, None]], format="csr")._is_array
+        assert block([[None, Fm], [Gm, None]], format="csr")._is_array
+        assert block([[Fm, Gm]], format="csr")._is_array
```

Comment:
I prefer using the attributes (both `.format` and `._is_array`) as I believe they are the intended mechanism to get this kind of information. But I could be wrong...

Reply:
I've opened https://github.com/scipy/scipy/issues/18921 to discuss `_is_array`, so happy to defer that for now.

I think the point about "why have the call specify `"csr"` if you don't check it" still stands. I think the check would make sense as is if the test was parameterized by `format`.

Reply:
The "csr" format is used because that format allows us to test the special as well as general cases in the code. I understand your desire to test the api rather than the code, and I'm fine with approaching testing that way. But my understanding is that the combinatorial explosion is hurting our existing tests (see test_base.py). With so many formats and so many different dtypes we end up with a lot of tests.

This test is not intended to test the `format` option. It is intended to test the return object. If we decide we should test all the formats, then we could easily decide to test all the dtypes as well -- even though this test isn't intended to test handling of dtype or format. What's the right approach to testing here?

Reply:
> The "csr" format is used because that format allows us to test the special as well as general cases in the code

Thanks for the additional context. My confusion here was that I didn't understand why the format was being specified. If the intent is to hit a certain codepath, that makes sense. I do think a comment explaining this could help so future editors know if/ when this should be changed.

I do think it would make sense to test the format somewhere (which I assume happens?), and that those tests *could* also cover the array/matrix types at the same time via instance checks. Looking locally it seems like there is coverage here so not a necessary change.